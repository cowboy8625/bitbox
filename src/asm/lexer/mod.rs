use std::{iter::Peekable, str::Chars};

pub fn lex(src: &str) -> Vec<Token> {
    Lexer::new(src).lex()
}

#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
pub struct Span {
    pub row_start: usize,
    pub row_end: usize,
    pub col_start: usize,
    pub col_end: usize,
    pub byte_start: usize,
    pub byte_end: usize,
}

impl From<(Span, Span)> for Span {
    fn from((start, end): (Span, Span)) -> Self {
        Self {
            row_start: start.row_start,
            row_end: end.row_end,
            col_start: start.col_start,
            col_end: end.col_end,
            byte_start: start.byte_start,
            byte_end: end.byte_end,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    KeywordPush,
    KeywordPop,
    KeywordLoad,
    KeywordAdd,
    KeywordSub,
    KeywordDiv,
    KeywordMul,
    KeywordEq,
    KeywordJne,
    KeywordInc,
    KeywordHult,
    KeywordPrintReg,
    KeywordAnd,
    KeywordOr,
    Number(u32),
    Identifier(String),
    Colon,
    Comma,
    Period,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    PercentSign,
    Ampersand,
    Equal,
    Delimiter,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

struct Lexer<'a> {
    src: Peekable<Chars<'a>>,
    span: Span,
    tokens: Vec<Token>,
    current: Option<char>,
    last_char: Option<char>,
}

impl<'a> Lexer<'a> {
    fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            span: Span::default(),
            tokens: Vec::new(),
            current: None,
            last_char: None,
        }
    }

    fn next(&mut self) -> Option<char> {
        self.last_char = self.current;
        self.current = self.src.next();
        match self.current {
            Some(c) if matches!(self.last_char, Some('\n')) => {
                self.span.row_end += 1;
                self.span.col_start = 0;
                self.span.col_end = 1;
                self.span.byte_end += 1;
                Some(c)
            }
            Some(c) => {
                self.span.col_end += 1;
                self.span.byte_end += 1;
                Some(c)
            }
            None => None,
        }
    }

    fn next_if(&mut self, pred: impl Fn(char) -> bool) -> Option<char> {
        match self.peek() {
            Some(c) if pred(c) => self.next(),
            _ => None,
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.src.peek().copied()
    }

    fn add_token(&mut self, kind: TokenKind) {
        let span = self.span();
        self.tokens.push(Token { kind, span });
    }

    fn add_token_with_span(&mut self, kind: TokenKind, span: Span) {
        self.tokens.push(Token { kind, span });
    }

    fn span(&mut self) -> Span {
        let span = self.span;
        self.span = Span {
            row_start: span.row_end,
            col_start: span.col_end,
            byte_start: span.byte_end,
            ..span
        };
        span
    }

    fn lex_number(&mut self, c: char) {
        let mut number = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_digit()) {
            number.push(c);
        }
        let span = self.span();
        self.add_token_with_span(
            TokenKind::Number(number.parse().expect("invalid number")),
            span,
        );
    }

    fn lex_identifier(&mut self, c: char) {
        let mut identifier = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_alphanumeric() || c == '_') {
            identifier.push(c);
        }

        let kind = match identifier.as_str() {
            "push" => TokenKind::KeywordPush,
            "pop" => TokenKind::KeywordPop,
            "inc" => TokenKind::KeywordInc,
            "load" => TokenKind::KeywordLoad,
            "add" => TokenKind::KeywordAdd,
            "sub" => TokenKind::KeywordSub,
            "div" => TokenKind::KeywordDiv,
            "mul" => TokenKind::KeywordMul,
            "eq" => TokenKind::KeywordEq,
            "jne" => TokenKind::KeywordJne,
            "hult" => TokenKind::KeywordHult,
            "printreg" => TokenKind::KeywordPrintReg,
            "and" => TokenKind::KeywordAnd,
            "or" => TokenKind::KeywordOr,
            _ => TokenKind::Identifier(identifier),
        };
        let span = self.span();
        self.add_token_with_span(kind, span);
    }

    fn comment(&mut self) {
        while let Some(_) = self.next_if(|c| c != '\n') {}
        self.span();
    }

    fn lex(mut self) -> Vec<Token> {
        while let Some(c) = self.next() {
            match c {
                '0'..='9' => self.lex_number(c),
                'a'..='z' | 'A'..='Z' | '_' => self.lex_identifier(c),
                ' ' | '\t' => {
                    self.span();
                }
                ';' => self.comment(),
                '=' => self.add_token(TokenKind::Equal),
                '\n' => self.add_token(TokenKind::Delimiter),
                '(' => self.add_token(TokenKind::LeftParen),
                ')' => self.add_token(TokenKind::RightParen),
                '{' => self.add_token(TokenKind::LeftBrace),
                '}' => self.add_token(TokenKind::RightBrace),
                '[' => self.add_token(TokenKind::LeftBracket),
                ']' => self.add_token(TokenKind::RightBracket),
                ',' => self.add_token(TokenKind::Comma),
                '.' => self.add_token(TokenKind::Period),
                ':' => self.add_token(TokenKind::Colon),
                '%' => self.add_token(TokenKind::PercentSign),
                '&' => self.add_token(TokenKind::Ampersand),
                _ => panic!("Unexpected character: {}", c),
            }
        }
        self.tokens
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    pub fn snapshot_lexing(input: &str) -> String {
        let tokens = lex(input);
        let mut tokens = std::collections::VecDeque::from(tokens);
        let mut output = String::new();
        for (row, line) in input.lines().enumerate() {
            output += line;
            output += "\n";
            while let Some(tok) = tokens.pop_front() {
                if tok.span.row_end != row {
                    tokens.push_front(tok);
                    break;
                }
                output += &" ".repeat(tok.span.col_start);
                output += &"^".repeat(tok.span.col_end.saturating_sub(tok.span.col_start));
                output += &format!(" {tok:?}");
                output += "\n"
            }
        }
        output
    }

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_lexing(contents));
                });
            }
        };
    }

    snapshot!(demo, "../samples/demo.asm");
    snapshot!(main, "../samples/main.asm");
    snapshot!(fib, "../samples/fib.asm");
}
