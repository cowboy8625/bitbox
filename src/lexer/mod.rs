mod lexer;
pub mod token;

#[cfg(test)]
mod test;

pub fn lex(src: &str) -> Vec<token::Token> {
    lexer::Lexer::new(src).collect()
}
