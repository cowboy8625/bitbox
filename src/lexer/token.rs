pub type Span = std::ops::Range<usize>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub span: Span,
}

impl Token {
    pub fn is_keyword(&self, keyword: Keyword) -> bool {
        matches!(self.kind, TokenKind::Keyword(ref k) if k == &keyword)
    }

    pub fn is_instruction(&self, instruction: Instruction) -> bool {
        matches!(self.kind, TokenKind::Instruction(ref i) if i == &instruction)
    }

    pub fn is_directive(&self, directive: Directive) -> bool {
        matches!(self.kind, TokenKind::Directive(ref d) if d == &directive)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TokenKind {
    Colon,
    Comma,
    Directive(Directive),
    Dot,
    Equals,
    Identifier,
    Instruction(Instruction),
    InvalidToken,
    Keyword(Keyword),
    LeftBrace,
    LeftBracket,
    LeftParen,
    Number,
    PathSeparator,
    Plus,
    RightBrace,
    RightBracket,
    RightParen,
    Semicolon,
    Star,
    String,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Instruction {
    Add,
    Call,
    Phi,
    Ret,
    Sub,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Directive {
    Len,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Keyword {
    Const,
    Import,
    Function,
    Public,
}
