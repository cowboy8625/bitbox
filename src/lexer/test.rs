use crate::lexer::lex;
use std::fmt::Write;

pub fn snapshot_lexing(input: &str) -> String {
    let tokens = lex(input);
    let mut tokens = std::collections::VecDeque::from(tokens);
    let mut output = String::new();
    let mut total = 0;
    for line in input.lines() {
        output += line;
        output += "\n";
        while let Some(tok) = tokens.pop_front() {
            if total + line.len() <= tok.get_span().start {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.get_span().start - total);
            output += &"^".repeat(tok.get_span().len());
            write!(&mut output, " {tok:?}").expect("failed to write()");
            output += "\n"
        }

        total += line.len() + 1;
    }

    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(binary, "../../snapshots/basic.bitbox");
snapshot!(import_function, "../../snapshots/import_function.bitbox");
