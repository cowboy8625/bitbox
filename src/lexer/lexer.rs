use super::token;
use crate::ast::{
    BBString, Builtin, Colon, Comma, Dot, Equals, Identifier, InvalidToken, LeftBrace, LeftParen,
    Number, Plus, RightBrace, RightParen, Semicolon,
};

type Token = Box<dyn token::Token>;
pub struct Lexer<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>,
    span: token::Span,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            chars: input.chars().peekable(),
            span: 0..0,
        }
    }

    fn next(&mut self) -> Option<char> {
        let value = self.chars.next();
        if value.is_some() {
            self.span.end += 1;
        }
        value
    }

    fn next_if(&mut self, predicate: impl Fn(char) -> bool) -> Option<char> {
        match self.chars.peek() {
            Some(value) if predicate(*value) => self.next(),
            _ => None,
        }
    }

    fn spanned(&mut self) -> token::Span {
        let span = self.span.clone();
        self.span = self.span.end..self.span.end;
        span
    }

    fn parse_number(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);

        while let Some(value) =
            self.next_if(|value| value.is_ascii_digit() || ['.', '_'].contains(&value))
        {
            lexeme.push(value);
        }

        token::create::<Number>(lexeme, self.spanned())
    }

    fn parse_identifier(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Identifier>(lexeme, self.spanned())
    }

    fn parse_builtin(&mut self) -> Token {
        let mut lexeme = String::from('@');
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Builtin>(lexeme, self.spanned())
    }

    fn parse_string(&mut self) -> Token {
        let mut lexeme = String::from('#');
        while let Some(value) = self.next() {
            lexeme.push(value);
            if lexeme.ends_with("\"#") {
                break;
            }
        }
        token::create::<BBString>(lexeme, self.spanned())
    }

    fn parse(&mut self) -> Option<Token> {
        match self.next() {
            Some(value @ '0'..='9') => Some(self.parse_number(value)),
            Some(value) if value.is_ascii_alphabetic() => Some(self.parse_identifier(value)),
            Some(value) if value.is_ascii_whitespace() => {
                self.spanned();
                self.parse()
            }
            Some('#') if self.chars.peek() == Some(&'"') => Some(self.parse_string()),
            Some('@') => Some(self.parse_builtin()),
            Some('+') => Some(token::create::<Plus>('+', self.spanned())),
            Some('(') => Some(token::create::<LeftParen>('(', self.spanned())),
            Some(')') => Some(token::create::<RightParen>(')', self.spanned())),
            Some('{') => Some(token::create::<LeftBrace>('(', self.spanned())),
            Some('}') => Some(token::create::<RightBrace>(')', self.spanned())),
            Some(':') => Some(token::create::<Colon>(':', self.spanned())),
            Some(';') => Some(token::create::<Semicolon>(';', self.spanned())),
            Some(',') => Some(token::create::<Comma>(',', self.spanned())),
            Some('=') => Some(token::create::<Equals>('=', self.spanned())),
            Some('.') => Some(token::create::<Dot>('.', self.spanned())),
            Some(value) => Some(token::create::<InvalidToken>(value, self.span.clone())),
            None => None,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        self.parse()
    }
}

#[test]
fn test_lexer() {
    let token = Lexer::new("#\"test\"#").parse().unwrap();
    assert_eq!(token.get_lexeme(), "#\"test\"#");
}
