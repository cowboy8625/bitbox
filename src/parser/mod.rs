#[cfg(test)]
mod test;
use crate::error::BitBoxError;
use crate::lexer::token::{self, Span, Token, TokenKind};
use crate::ssa::{self, IntoSsaType};

enum TopLevel {
    Function(ssa::Function),
    Import(ssa::Import),
    Constant(ssa::Constant),
}

pub struct Parser {
    stream: std::iter::Peekable<std::vec::IntoIter<Token>>,
}

// Helpers
impl Parser {
    fn parse_visibility(&mut self) -> ssa::Visibility {
        self.stream
            .next_if(|token| token.is_keyword(token::Keyword::Public))
            .map(|_| ssa::Visibility::Public)
            .unwrap_or_default()
    }

    fn consume(&mut self, expected: TokenKind) -> Result<Token, BitBoxError> {
        match self.stream.next() {
            Some(actual) if actual.kind == expected => Ok(actual),
            Some(actual) => Err(BitBoxError::UnexpectedToken { expected, actual }),
            None => Err(BitBoxError::UnexpectedEndOfStream),
        }
    }

    fn is_peek_a(&mut self, kind: TokenKind) -> bool {
        matches!(self.stream.peek(), Some(token) if token.kind == kind)
    }

    fn end_of_stream(&mut self) -> bool {
        self.stream.peek().is_none()
    }

    fn next(&mut self) -> Result<Token, BitBoxError> {
        let Some(token) = self.stream.next() else {
            return Err(BitBoxError::UnexpectedEndOfStream);
        };
        Ok(token)
    }
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            stream: tokens.into_iter().peekable(),
        }
    }

    pub fn parse(&mut self) -> Result<ssa::Program, BitBoxError> {
        let mut imports = vec![];
        let mut functions = vec![];
        let mut constants = vec![];

        while !self.end_of_stream() {
            match self.parse_top_level()? {
                TopLevel::Import(import) => imports.push(import),
                TopLevel::Function(func) => functions.push(func),
                TopLevel::Constant(constant) => constants.push(constant),
            }
        }

        Ok(ssa::Program {
            functions,
            imports,
            constants,
        })
    }

    fn parse_top_level(&mut self) -> Result<TopLevel, BitBoxError> {
        let visibility = self.parse_visibility();
        if self.is_peek_a(TokenKind::Keyword(token::Keyword::Function)) {
            Ok(TopLevel::Function(self.parse_function(visibility)?))
        } else if self.is_peek_a(TokenKind::Keyword(token::Keyword::Import)) {
            Ok(TopLevel::Import(self.parse_import()?))
        } else if self.is_peek_a(TokenKind::Keyword(token::Keyword::Const)) {
            Ok(TopLevel::Constant(self.parse_constant()?))
        } else {
            let tok = self.next()?;
            Err(BitBoxError::ExpectedTopLevelItem(tok))
        }
    }

    fn parse_function(
        &mut self,
        visibility: ssa::Visibility,
    ) -> Result<ssa::Function, BitBoxError> {
        self.consume(TokenKind::Keyword(token::Keyword::Function))?;
        let func_name = self.consume(TokenKind::Identifier)?;
        let params = self.parse_function_params()?;
        let return_type = self.parse_type()?;
        let blocks = self.parse_function_block()?;

        Ok(ssa::Function {
            visibility,
            name: func_name.lexeme,
            params,
            return_type,
            blocks,
        })
    }

    fn parse_function_params(&mut self) -> Result<Vec<ssa::Variable>, BitBoxError> {
        self.consume(TokenKind::LeftParen)?;
        let mut params = vec![];

        let mut version = 0;
        while !self.end_of_stream() && !self.is_peek_a(TokenKind::RightParen) {
            let name = self.consume(TokenKind::Identifier)?;
            self.consume(TokenKind::Colon)?;
            let ty = self.parse_type()?;

            let param = ssa::Variable { name, ty, version };
            params.push(param);
            if !self.is_peek_a(TokenKind::Comma) {
                break;
            }
            self.consume(TokenKind::Comma)?;
            version += 1;
        }
        self.consume(TokenKind::RightParen)?;
        Ok(params)
    }

    fn parse_function_block(&mut self) -> Result<Vec<ssa::BasicBlock>, BitBoxError> {
        let mut blocks = vec![];
        self.consume(TokenKind::LeftBrace)?;
        while !self.end_of_stream() && !self.is_peek_a(TokenKind::RightBrace) {
            let block = self.parse_basic_block()?;
            blocks.push(block);
        }
        self.consume(TokenKind::RightBrace)?;
        Ok(blocks)
    }

    fn parse_basic_block(&mut self) -> Result<ssa::BasicBlock, BitBoxError> {
        let mut instructions = vec![];

        while !self.end_of_stream() && !self.is_peek_a(TokenKind::RightBrace) {
            let Some(instruction) = self.parse_instruction()? else {
                break;
            };
            instructions.push(instruction);
        }

        Ok(ssa::BasicBlock {
            id: 0,
            instructions,
            // TODO: Add predecessors and successors
            successors: vec![],
            predecessors: vec![],
        })
    }

    fn parse_instruction(&mut self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        let tok = self.next()?;
        let Token {
            kind: TokenKind::Instruction(instruction),
            lexeme,
            span,
        } = tok
        else {
            return Err(BitBoxError::InvalidInstruction(tok));
        };
        match instruction {
            token::Instruction::Ret => self.parse_return(),
            token::Instruction::Add => self.parse_add(),
            token::Instruction::Sub => self.parse_sub(),
            token::Instruction::Call => self.parse_call(),
            token::Instruction::Phi => self.parse_phi(),
        }
    }

    fn parse_arguments(&mut self) -> Result<Vec<ssa::Operand>, BitBoxError> {
        self.consume(TokenKind::LeftParen)?;
        let mut args = vec![];
        while !self.end_of_stream() && !self.is_peek_a(TokenKind::RightParen) {
            let arg = self.parse_operand()?;
            args.push(arg);
            if !self.is_peek_a(TokenKind::Comma) {
                break;
            }
            self.consume(TokenKind::Comma)?;
        }
        self.consume(TokenKind::RightParen)?;
        Ok(args)
    }

    fn parse_operand(&mut self) -> Result<ssa::Operand, BitBoxError> {
        if self.is_peek_a(TokenKind::Number) {
            let tok = self.consume(TokenKind::Number)?;
            return Ok(ssa::Operand::Constant(tok));
        }
        let tok = self.consume(TokenKind::Identifier)?;
        Ok(ssa::Operand::Variable(tok))
    }

    fn parse_import_function_params(&mut self) -> Result<Vec<ssa::Type>, BitBoxError> {
        let mut params = vec![];
        self.consume(TokenKind::LeftParen)?;
        while !self.end_of_stream() {
            let ty = self.parse_type()?;
            params.push(ty);
            if !self.is_peek_a(TokenKind::Comma) {
                break;
            }
            self.consume(TokenKind::Comma)?;
        }
        self.consume(TokenKind::RightParen)?;
        Ok(params)
    }

    fn parse_import(&mut self) -> Result<ssa::Import, BitBoxError> {
        self.consume(TokenKind::Keyword(token::Keyword::Import))?;
        self.consume(TokenKind::Keyword(token::Keyword::Function))?;
        let module_name = self.consume(TokenKind::Identifier)?;
        self.consume(TokenKind::PathSeparator)?;
        let name = self.consume(TokenKind::Identifier)?;
        let params = self.parse_import_function_params()?;
        let return_type = self.parse_type()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(ssa::Import::Function(ssa::FunctionSpec {
            module_name,
            name,
            params,
            return_type,
        }))
    }

    fn parse_constant_value(&mut self) -> Result<ssa::ConstantValue, BitBoxError> {
        let tok = self.next()?;
        match tok.kind {
            TokenKind::String => Ok(ssa::ConstantValue::String(tok)),
            TokenKind::Directive(directive) => match directive {
                token::Directive::Len => {
                    let value = self.consume(TokenKind::Identifier)?;
                    Ok(ssa::ConstantValue::Directive(ssa::Directive::Len(value)))
                }
            },
            _ => Err(BitBoxError::InvalidContantValue(tok)),
        }
    }

    fn parse_constant(&mut self) -> Result<ssa::Constant, BitBoxError> {
        self.consume(TokenKind::Keyword(token::Keyword::Const))?;
        let name = self.consume(TokenKind::Identifier)?;
        self.consume(TokenKind::Colon)?;
        let ty = self.parse_type()?;
        self.consume(TokenKind::Equals)?;
        let value = self.parse_constant_value()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(ssa::Constant { name, ty, value })
    }

    fn parse_type(&mut self) -> Result<ssa::Type, BitBoxError> {
        let tok = self.next()?;
        match tok.kind {
            TokenKind::Identifier => tok
                .into_ssa_type()
                .map_err(|tok| BitBoxError::ExpectedType(tok)),
            TokenKind::Star => {
                let ty = self.parse_type()?;
                Ok(ssa::Type::Pointer(Box::new(ty)))
            }
            TokenKind::LeftBracket => {
                let count = self.consume(TokenKind::Number)?;
                self.consume(TokenKind::Semicolon)?;
                let ty = self.parse_type()?;
                self.consume(TokenKind::RightBracket)?;
                Ok(ssa::Type::Array(
                    count.lexeme.parse().unwrap(),
                    Box::new(ty),
                ))
            }
            _ => Err(BitBoxError::ExpectedType(tok)),
        }
    }

    fn parse_return(&mut self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        let ty = self.parse_type()?;
        self.consume(TokenKind::Colon)?;
        let value = self.parse_operand()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(Some(ssa::Instruction::Return(ty, value)))
    }

    fn parse_add(&mut self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        let ty = self.parse_type()?;
        self.consume(TokenKind::Colon)?;
        let name = self.consume(TokenKind::Identifier)?;
        let des = ssa::Variable {
            name,
            ty,
            version: 0,
        };
        self.consume(TokenKind::Comma)?;
        let lhs = self.parse_operand()?;
        self.consume(TokenKind::Comma)?;
        let rhs = self.parse_operand()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(Some(ssa::Instruction::Add(des, lhs, rhs)))
    }

    fn parse_sub(&mut self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        let ty = self.parse_type()?;
        self.consume(TokenKind::Colon)?;
        let name = self.consume(TokenKind::Identifier)?;
        let des = ssa::Variable {
            name,
            ty,
            version: 0,
        };
        self.consume(TokenKind::Comma)?;
        let lhs = self.parse_operand()?;
        self.consume(TokenKind::Comma)?;
        let rhs = self.parse_operand()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(Some(ssa::Instruction::Sub(des, lhs, rhs)))
    }

    fn parse_call(&mut self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        let ty = self.parse_type()?;
        self.consume(TokenKind::Colon)?;
        let result_name = self.consume(TokenKind::Identifier)?;
        let des = ssa::Variable {
            name: result_name,
            ty,
            version: 0,
        };
        let name = self.consume(TokenKind::Identifier)?;
        let arguments = self.parse_arguments()?;
        self.consume(TokenKind::Semicolon)?;
        Ok(Some(ssa::Instruction::Call(des, name, arguments)))
    }

    fn parse_phi(&self) -> Result<Option<ssa::Instruction>, BitBoxError> {
        todo!("implement phi instruciton parser")
    }
}
