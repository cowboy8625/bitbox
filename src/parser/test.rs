use crate::lexer::lex;
use crate::parser;

pub fn snapshot_parsing(input: &str) -> String {
    let tokens = lex(input);
    let program = parser::Parser::new(tokens).parse();
    format!("{:#?}", program)
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_parsing(contents));
            });
        }
    };
}

snapshot!(binary, "../../snapshots/basic.bitbox");

