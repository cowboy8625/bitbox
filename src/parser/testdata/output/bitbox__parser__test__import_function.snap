---
source: src/parser/test.rs
expression: snapshot_parsing(contents)
---
Ok(
    Program {
        imports: [
            Function(
                FunctionSpec {
                    module_name: Token {
                        kind: Identifier,
                        lexeme: "core",
                        span: 16..20,
                    },
                    name: Token {
                        kind: Identifier,
                        lexeme: "write",
                        span: 22..27,
                    },
                    params: [
                        Signed(
                            32,
                        ),
                        Signed(
                            32,
                        ),
                    ],
                    return_type: Signed(
                        32,
                    ),
                },
            ),
        ],
        constants: [
            Constant {
                name: Token {
                    kind: Identifier,
                    lexeme: "value_ptr",
                    span: 50..59,
                },
                ty: Pointer(
                    Array(
                        13,
                        Unsigned(
                            8,
                        ),
                    ),
                ),
                value: String(
                    Token {
                        kind: String,
                        lexeme: "Hello World\n",
                        span: 74..91,
                    },
                ),
            },
            Constant {
                name: Token {
                    kind: Identifier,
                    lexeme: "value_len",
                    span: 99..108,
                },
                ty: Unsigned(
                    32,
                ),
                value: Directive(
                    Len(
                        Token {
                            kind: Identifier,
                            lexeme: "value_ptr",
                            span: 122..131,
                        },
                    ),
                ),
            },
        ],
        functions: [
            Function {
                visibility: Public,
                name: "main",
                params: [],
                return_type: Signed(
                    32,
                ),
                blocks: [
                    BasicBlock {
                        id: 0,
                        instructions: [
                            Call(
                                Variable {
                                    name: Token {
                                        kind: Identifier,
                                        lexeme: "result",
                                        span: 179..185,
                                    },
                                    ty: Signed(
                                        32,
                                    ),
                                    version: 0,
                                },
                                Token {
                                    kind: Identifier,
                                    lexeme: "write",
                                    span: 186..191,
                                },
                                [
                                    Variable(
                                        Token {
                                            kind: Identifier,
                                            lexeme: "value_ptr",
                                            span: 192..201,
                                        },
                                    ),
                                    Variable(
                                        Token {
                                            kind: Identifier,
                                            lexeme: "value_len",
                                            span: 203..212,
                                        },
                                    ),
                                ],
                            ),
                            Return(
                                Signed(
                                    32,
                                ),
                                Variable(
                                    Token {
                                        kind: Identifier,
                                        lexeme: "result",
                                        span: 230..236,
                                    },
                                ),
                            ),
                        ],
                        successors: [],
                        predecessors: [],
                    },
                ],
            },
        ],
    },
)
